micro-rdk/src/common/data_manager.rs

```diff
@@ -142,59 +149,32 @@ where
         intervals
     }
 
-    pub async fn run(&mut self) -> Result<(), DataManagerError> {
+    pub async fn data_collection_task(&mut self) -> Result<(), DataManagerError> {
         let mut loop_counter: u64 = 0;
         loop {
-            self.run_inner(loop_counter)?;
+            self.collect_data_inner(loop_counter).await?;
             loop_counter += 1;
             Timer::after(self.min_interval).await;
         }
     }
 
-    fn run_inner(&mut self, loop_counter: u64) -> Result<(), DataManagerError> {
+    pub async fn collect_data_inner(&mut self, loop_counter: u64) -> Result<(), DataManagerError> {
         let min_interval_ms = self.min_interval_ms();
-        if (loop_counter % (self.sync_interval_ms() / min_interval_ms)) == 0 && (loop_counter != 0)
-        {
-            self.sync()?;
-        }
         for interval in self.collection_intervals() {
             if loop_counter % (interval / min_interval_ms) == 0 {
-                self.collect_and_store_readings(interval)?;
-            }
-        }
-        Ok(())
-    }
-
-    fn sync(&mut self) -> Result<(), DataManagerError> {
-        for collector_key in self.collectors.iter().map(|c| c.resource_method_key()) {
-            // TODO: check for internet access before attempting to read from store
-            let mut readings_to_upload: Vec<BytesMut> = vec![];
-            let mut reader = self.store.get_reader(&collector_key)?;
-            loop {
-                match reader.read_next_message() {
-                    Ok(msg) => {
-                        if msg.is_empty() {
-                            break;
-                        }
-                        readings_to_upload.push(msg);
-                    }
-                    Err(err) => return Err(err.into()),
-                };
+                self.collect_and_store_readings(interval).await?;
             }
-            // TODO: implement actual upload logic here, will likely have to change struct
-            // and make this function async. Also flushing below should only occur on upload success
-            reader.flush();
         }
         Ok(())
     }
 
-    fn collect_and_store_readings(
+    async fn collect_and_store_readings(
         &mut self,
         time_interval_ms: u64,
     ) -> Result<(), DataManagerError> {
         for (collector_key, reading) in self.collect_readings_for_interval(time_interval_ms)? {
-            self.store
-                .write_message(&collector_key, reading, WriteMode::OverwriteOldest)?
+            let mut store_guard = self.store.lock().await;
```

Comment:
Do you want to lock and unlock for each collector key? Or lock once and do all collector keys under the lock?

Reply:
I'm inclined to lock for each collector key so that conceivably a non-interfering write can happen in between unrelated uploads?